test/e2e/istio-csr_test.go

```
@@ -0,0 +1,252 @@
+														ExpirationSeconds: ptr.To(int64(3600)),
+														Path:              "token",
+													},
+												},
+											},
+										},
+									},
+								},
+								{
+									Name: "root-secret",
+									VolumeSource: corev1.VolumeSource{
+										Secret: &corev1.SecretVolumeSource{
+											SecretName: "istiod-tls",
+										},
+									},
+								},
+								{
+									Name: "proto",
+									VolumeSource: corev1.VolumeSource{
+										ConfigMap: &corev1.ConfigMapVolumeSource{
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: "proto-cm",
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+			_, err = clientset.BatchV1().Jobs(ns.Name).Create(context.TODO(), job, metav1.CreateOptions{})
+			Expect(err).Should(BeNil())
+			defer clientset.BatchV1().Jobs(ns.Name).Delete(ctx, job.Name, metav1.DeleteOptions{})
+
+			By("waiting for the job to be completed")
+			err = pollTillJobCompleted(ctx, clientset, ns.Name, "grpcurl-job")
+			Expect(err).Should(BeNil())
+
+			By("fetching logs of the grpcurl job")
+			pods, err := clientset.CoreV1().Pods(ns.Name).List(context.TODO(), metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app=%s", grpcAppName),
+			})
+			Expect(err).Should(BeNil())
+
+			By("reading logs of the grpcurl job")
+			for _, pod := range pods.Items {
+				req := clientset.CoreV1().Pods(ns.Name).GetLogs(pod.Name, &corev1.PodLogOptions{})
+				logs, err := req.Stream(context.TODO())
+				Expect(err).Should(BeNil())
+
+				defer logs.Close()
+
+				logData, err := io.ReadAll(logs)
+				Expect(err).Should(BeNil())
+
+				var entry LogEntry
+				err = json.Unmarshal(logData, &entry)
+				Expect(err).Should(BeNil())
+
+				By("validating each certificate")
+				for _, certPEM := range entry.CertChain {
+					err = library.ValidateCertificate(certPEM, "my-selfsigned-ca")
+					Expect(err).Should(BeNil())
+				}
```

since these are Pods from a Job, under some circumstance we might see a pod that had failed, the Job would retry in that case.

Hence, make this loop such that we can get a "atleast once" guarantee. Also, ignore errors otherwise.

---

For retry, there should be retry attempt or time limit?

---

set `.spec.completions = 1` and `.spec.backoffLimit = N`, N can be a reasonable retry limit like 10. 

https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures

---

Filter out the pod loop by looking only into pods with status Completed.


what I mean is:
```

desiredPodName := ""
for _, pod := range pods.Items {
    if pod.Status.Phase == "Succeeded" {
            desiredPodName = pod.Name
    }
}

Expect(desiredPodName).ShouldNot(BeEmpty())


req := clientset.CoreV1().Pods(ns.Name).GetLogs(desiredPodName, &corev1.PodLogOptions{})
// ...
```

---

Done.